safekeeper/src/send_interpreted_wal.rs

```diff
@@ -103,20 +335,181 @@ impl<IO: AsyncRead + AsyncWrite + Unpin> InterpretedWalSender<'_, IO> {
+                    match &mut *self.state.write().unwrap() {
+                        InterpretedWalReaderState::Running { current_position, .. } => {
+                            *current_position = max_next_record_lsn;
+                        },
+                        InterpretedWalReaderState::Done => {
+                            unreachable!()
+                        }
+                    }
+
+                    // Send interpreted records downstream. Anything that has already been seen
+                    // by a shard is filtered out.
+                    let mut shard_senders_to_remove = Vec::new();
+                    for (shard, states) in &mut self.shard_senders {
+                        for state in states {
+                            if max_next_record_lsn <= state.next_record_lsn {
+                                continue;
+                            }
+
+                            let shard_sender_id = ShardSenderId::new(*shard, state.sender_id);
+                            let records = records_by_sender.remove(&shard_sender_id).unwrap_or_default();
 
-                    tx.send(Batch {wal_end_lsn, available_wal_end_lsn, records: batch}).await.unwrap();
+                            let batch = InterpretedWalRecords {
+                                records,
+                                next_record_lsn: Some(max_next_record_lsn),
+                            };
+
+                            let res = state.tx.send(Batch {
+                                wal_end_lsn,
+                                available_wal_end_lsn,
+                                records: batch,
+                            }).await;
+
+                            if res.is_err() {
+                                shard_senders_to_remove.push(shard_sender_id);
+                            } else {
+                                state.next_record_lsn = max_next_record_lsn;
+                            }
+                        }
+                    }
+
+                    // Clean up any shard senders that have dropped out.
+                    // This is inefficient, but such events are rare (connection to PS termination)
+                    // and the number of subscriptions on the same shards very small (only one
+                    // for the steady state).
+                    for to_remove in shard_senders_to_remove {
+                        let shard_senders = self.shard_senders.get_mut(&to_remove.shard()).expect("saw it above");
+
+                        let mut remove_at = None;
+                        for (idx, shard_sender) in shard_senders.iter().enumerate() {
+                            let crnt_shard_sender_id = ShardSenderId::new(to_remove.shard(), shard_sender.sender_id);
+                            if crnt_shard_sender_id == to_remove {
+                                remove_at = Some(idx);
+                                break;
+                            }
+                        }
```

Comment:
nit: this can be replaced by `if let Some(i) = shard_senders.iter().position(|s| s.sender_id == to_remove.sender_id) { ... }`.

Reply:
https://github.com/neondatabase/neon/pull/10190/commits/1713a425d85d69b7e49a3b2687ee7559517d4545

I also realised we don't remove the entry from the hash map if the vec is now empty. Fixed in https://github.com/neondatabase/neon/pull/10190/commits/248817f01175aefee82b520d058d58fab006fe2e